cni/pkg/nodeagent/cni-watcher.go

```diff
@@ -0,0 +1,182 @@
+			log.Errorf("Error running CNI listener server: %v", err)
+		}
+	}()
+
+	go func() {
+		<-s.ctx.Done()
+		if err := s.cniListenServer.Close(); err != nil {
+			log.Errorf("CNI listen server terminated with error: %v", err)
+		} else {
+			log.Debug("CNI listen server terminated")
+		}
+	}()
+
+	return nil
+}
+
+func (s *CniPluginServer) handleAddEvent(w http.ResponseWriter, req *http.Request) {
+	if req.Body == nil {
+		log.Error("empty request body")
+		http.Error(w, "empty request body", http.StatusBadRequest)
+		return
+	}
+	defer req.Body.Close()
+	data, err := io.ReadAll(req.Body)
+	if err != nil {
+		log.Errorf("Failed to read event report from cni plugin: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	msg, err := processAddEvent(data)
+	if err != nil {
+		log.Errorf("Failed to process CNI event payload: %v", err)
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	if err := s.ReconcileCNIAddEvent(req.Context(), msg); err != nil {
+		log.Errorf("Failed to handle add event: %v", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+func processAddEvent(body []byte) (CNIPluginAddEvent, error) {
+	var msg CNIPluginAddEvent
+	err := json.Unmarshal(body, &msg)
+	if err != nil {
+		log.Errorf("Failed to unmarshal CNI plugin event: %v", err)
+		return msg, err
+	}
+
+	log.Debugf("Deserialized CNI plugin event: %+v", msg)
+	return msg, nil
+}
+
+func (s *CniPluginServer) ReconcileCNIAddEvent(ctx context.Context, addCmd CNIPluginAddEvent) error {
+	log := log.WithLabels("cni-event", addCmd)
+
+	// The CNI node plugin should have already checked the pod with this
+	// exact same function before forwarding us the event, but we have to invoke the K8S client anyway,
+	// so to be safe we check it again here to make sure we get the same result
+	ambientPod, err := s.handlers.GetPodIfAmbient(addCmd.PodName, addCmd.PodNamespace)
+	if err != nil {
+		return err
+	}
+
+	if ambientPod != nil {
+		log.Debugf("Pod: %s in ns: %s is enabled for ambient, adding to mesh. ", addCmd.PodName, addCmd.PodNamespace)
+
+		var podIps []netip.Addr
+		for _, configuredPodIPs := range addCmd.IPs {
+			// net.ip is implicitly convertible to netip as slice
+			ip, _ := netip.AddrFromSlice(configuredPodIPs.Address.IP)
+			// We ignore the mask of the IPNet - it's fine if the IPNet defines
+			// a block grant of addresses, we just need one for checking routes.
+			podIps = append(podIps, ip)
+		}
+		// Note that we use the IP info from the CNI plugin here - the Pod struct as reported by K8S doesn't have this info
+		// yet (because the K8S control plane doesn't), so it will be empty there.
+		err := s.dataplane.AddPodToMesh(ctx, ambientPod, podIps, addCmd.Netns)
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("got an unexpected event for pod %s in namespace %s! This is a plugin bug", addCmd.PodName, addCmd.PodNamespace)
```

Comment:
The above comment mentions this:

```
	// The CNI node plugin should have already checked the pod with this
	// exact same function before forwarding us the event, but we have to invoke the K8S client anyway,
	// so to be safe we check it again here to make sure we get the same result
```

This check is arguably redundant/paranoia - we cannot get here unless the CNI plugin on our node has already successfully performed the exact same check, and forwarded the event to us.

In our testing it has never been observed, I am fine with dropping it here, the plugin does and should already handle this.

Reply:
Then our options are to trust one of the caches, or block here until this one is consistent, or we decide the other is. Do you have a preference?

Reply:
The plugin isn't a cache its doing a GET to the API server in a loop fwiw.Seems like we have to wait for it here, otherwise won't the CNI plugin crashloop?

Reply:
Ah that's right, because of the extra caching machinery on this end.

So the plugin, if it is right, will always be right, and we should be able to retry here until we reach consensus.